### ISSUE TITLE
`json_parse` can crash without good reason

### ISSUE THREAD

**Issue by sanbox-irl** [YoYoGames/GameMaker-Bugs#6130]
### Description

This error happens **infrequently and seemingly randomly**:
We get no error message except `JSON parse error`, and we parse like this `json_parse(json_txt);`

The input string was:

```
{"type":"world_fact_response","content":{"type":"float","content":1.0}}
```

I debug out the following:

```
str_len = `71`

0 = `{`
1 = `\"`
2 = `t`
3 = `y`
4 = `p`
5 = `e`
6 = `\"`
7 = `:`
8 = `\"`
9 = `w`
10 = `o`
11 = `r`
12 = `l`
13 = `d`
14 = `_`
15 = `f`
16 = `a`
17 = `c`
18 = `t`
19 = `_`
20 = `r`
21 = `e`
22 = `s`
23 = `p`
24 = `o`
25 = `n`
26 = `s`
27 = `e`
28 = `\"`
29 = `,`
30 = `\"`
31 = `c`
32 = `o`
33 = `n`
34 = `t`
35 = `e`
36 = `n`
37 = `t`
38 = `\"`
39 = `:`
40 = `{`
41 = `\"`
42 = `t`
43 = `y`
44 = `p`
45 = `e`
46 = `\"`
47 = `:`
48 = `\"`
49 = `f`
50 = `l`
51 = `o`
52 = `a`
53 = `t`
54 = `\"`
55 = `,`
56 = `\"`
57 = `c`
58 = `o`
59 = `n`
60 = `t`
61 = `e`
62 = `n`
63 = `t`
64 = `\"`
65 = `:`
66 = `1`
67 = `.`
68 = `0`
69 = `}`
70 = `}`
```

Everything here looks valid and normal as far as I can tell

### Steps To Reproduce

In Fields of Mistria, this can happen occasionally, but seemingly randomly.

### How reliably can you recreate this issue using your steps above?

Rarely

### Which version of GameMaker are you reporting this issue for?

2024.4.1 (Monthly)

### Which platform(s) are you seeing the problem on?

Windows

### Contact Us Package Attached?

- [X] I have attached my Contact Us Package

### Sample Project Added?

- [ ] I have included a small sample project

**Comment by jackerley** [comment]
Duplicate of https://github.com/YoYoGames/GameMaker-Bugs/issues/5986

**Comment by sanbox-irl** [comment]
@jackerley this is not correct -- we have this crashing on similar strings without numbers at all, and the error message we get back says nothing about numbers.

Here's another example (in fact, this is the most common failure case):

```
{"type":"actions","content":[]}
```

That issue might be related, but I doubt it -- the above string will parse just fine *most* of the time, but sometimes fails.

**Comment by rwkay** [comment]
This was happening on monthly not on 2024.600 the locale issue was only on 2024.600 not on 2024.4.1

**Comment by jackerley** [comment]
Hi could you add me to your github repo so I can check this out? I don't believe I have access atm.

Git Username jackerley

Thanks,
Fritz

**Comment by sanbox-irl** [comment]
@jackerley you've been added! The repository is called SwordAndField. Instructions for running the game for ya'll is the following:

- Clone the repo
- Run `.build.ps` (or `.build.sh` on macOS (only supports arm builds))
- You can now compile the game in GameMaker

We use a custom build tool called `adam` which comes in the project -- so you could instead do `./tools/bin/adam.exe run` to run the project, but it can require some setup, so it might not be worth your time.

The offending source of the bug is T2R.gml:579, the function `parse_t2_output`. You can see the try-catch and our logging there.

**Comment by jackerley** [comment]
@sanbox-irl I don't seem to have received an invite and it doesn't show when I browse on github, seems strange

**Comment by sanbox-irl** [comment]
@jackerley should be good to go now

**Comment by sanbox-irl** [comment]
Also not sure if this is your department, but if you could check out <https://github.com/YoYoGames/GameMaker-Bugs/issues/6200> while you're in there, that'd be a godsend. We want to do optimizations before EA release in August and we're pretty much just using traces with times to do it, which is obviously rough going

**Comment by jackerley** [comment]
Thanks, I have the project up and running now, how do I go about reproducing the issue?

**Comment by sanbox-irl** [comment]
> Thanks, I have the project up and running now, how do I go about reproducing the issue?

It's quite intermittent, but common in large numbers of playtesters. I think out of tens of thousands of playing event, we have ~150 incidents of it. This makes it not a massive deal, but it's by far our biggest bug. Reproducing a save's exact behavior is insufficient to make it happen.

I told Russell, but we're also plagued by segfaults from GM's side, whose origin we haven't managed to track down, and I suppose it's possible this is the result of that. Anything's really possible in fun time UB land. I wanted to bring that up if that answer makes sense to you.

Otherwise, I will try to fuzz a reproduction of this bug for you, but likely not till Monday

**Comment by jackerley** [comment]
Hi @sanbox-irl did you ever manage to get a reproduction of this sorted?

**Comment by sanbox-irl** [comment]
@jackerley Not yet, and even now, having just released in EA, we are getting reports of this. Extremely odd, since previously this was a string given by a buffer (which I audited for any mem safety problems, and found none), but now we've switched to a shared buffer managed by GM (a la, `buffer_create` + `buffer_get_address`). So the problem is likely not to do with the surrounding code for getting a string. I will be pursuing this again tomorrow...just as soon as we're done celebrating release 😉

**Comment by jackerley** [comment]
Hi @sanbox-irl any update?

**Comment by sanbox-irl** [comment]
@jackerley Unfortunately no -- switching to a buffer for the DLL had to effect on incidence rate (which means I believe we can discount the DLL's involvement as meaningful). We simply have radically reduced the amount of JSON parsing which we do in the game, because it seems to fail eventually after repeated usage -- but it still crashes.

The only extra thing I can say here is that technically `json_parse` doesn't directly crash, which I didn't know above -- instead it's just returning `undefined` (or at least appears to -- this is all in a try/catch).

Here's the whole code path:

```
var json_txt = buffer_peek(self.bridge, 0, buffer_string);

var raw_output;
try {
	raw_output = json_parse(json_txt);
	switch raw_output.type { /* snip */ } // we actually crash on `raw_output.type` because `raw_output` is undefined
} catch(e) {
	error("raw_input:\n`{}`", json_txt);
	error("we parsed as: {}..type={}", raw_output, typeof(raw_output));

	CRASH("Json parse error: {}", e.longMessage);
};
```

The output we get via sentry is:

```
Json parse error: ERROR in action number 1 of Step Event0 for object Game: I32 argument is undefined

raw_input:
`{"type":"schedule_state","content":{"point":"inn/South Table Left End","time":45120,"actions":[{"type":"animation","content":{"npc":"errol","animation":"idle"}},{"type":"can_talk","content":"errol"}]}}`

we parsed as: undefined..type=undefined
```

We don't have particularly more to go on than this -- thinking of the Crow Country JSON-parse error, I did check into the locales of people who have reported this bug (even though the above string has no decimals, and therefore shouldn't have even the possibility of a weird locale error), but users who have experienced this seem to come from all over, so I don't think it's particular to a distinguished class of users.

I suppose the only extra clue I can now offer is that `json_parse` is returning `undefined` and we aren't sure in what scenario it is supposed to do that (the docs seem to say it never will)

**Comment by jackerley** [comment]
@sanbox-irl  Hmm, that's interesting, so it's failing on an input that is valid, something must be going wrong internally. Checking the code json_parse will fail and return undefined if it's passed NULL or if the parsing fails. I've put a test together that batters json_parse with that mixed with other valid json strings and I can't get it to fail. Is there any chance that you might be running out of memory at the point at which it's failing?

**Comment by sanbox-irl** [comment]
@jackerley it could *possibly* be a memory issue, but wouldn't that throw a memory allocation bug, rather than just silently return an undefined? Moreover, looking over the sentry reports, there are definitely people with a lot of memory but it's not crazy -- something around 4gigs allocated at the time of crash. That could be OOM if someone has a lot else going or a low-end computer, but doesn't seem particularly crazy to me.

Is there a scenario where `json_parse` can return `undefined` *without* also raising an exception? Because that does appear to be what's happening here -- the actual line that crashes is the next line where we try to read the `undefined`

**Comment by shichen85** [comment]
> [@jackerley](https://github.com/jackerley) it could _possibly_ be a memory issue, but wouldn't that throw a memory allocation bug, rather than just silently return an undefined? Moreover, looking over the sentry reports, there are definitely people with a lot of memory but it's not crazy -- something around 4gigs allocated at the time of crash. That could be OOM if someone has a lot else going or a low-end computer, but doesn't seem particularly crazy to me.
> 
> Is there a scenario where `json_parse` can return `undefined` _without_ also raising an exception? Because that does appear to be what's happening here -- the actual line that crashes is the next line where we try to read the `undefined`

@sanbox-irl  Sorry to necro this one, but did you find any mitigations for this issue? We are seeing this error message too in some players, and it does smell like memory issue, which is so hard to debug!

**Comment by sanbox-irl** [comment]
@shichen85 we generally have changed how we program GML to allocate as few objects as possible. particularly, we were decoded some JSON every frame at least, and that's been altered to be as rarely as possible. overall, these actions have reduced incidence, but it's still a bug source. i haven't heard from YYG on this in months
